stdlib/pqueue.ml

```diff
@@ -0,0 +1,223 @@
+(*                                                                        *)
+(*   Copyright 2023 CNRS                                                  *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Priority queues over ordered elements.
+
+   We choose to have polymorphic elements here, so that we can later
+   instantiate [MakePoly] on polymorphic pairs (see functor [Make] below).
+*)
+
+module MakePoly(E: sig
+    type 'a t
+    val compare: 'a t -> 'a t -> int
+end) =
+  struct
+
+    type 'a elt = 'a E.t
+
+    (* Our priority queues are implemented using the standard "min heap"
+       data structure, a dynamic array representing a binary tree. *)
+    type 'a t = 'a E.t Dynarray.t
+
+    let create =
+      Dynarray.create
+
+    let length =
+      Dynarray.length
+
+    let is_empty h =
+      length h = 0
+
+    let clear =
+      Dynarray.clear
+
+    (* The node at index [i] has children nodes at indices [2 * i + 1]
+       and [2 * i + 2] -- if they are valid indices in the dynarray. *)
+    let left_child i = 2 * i + 1
+    let right_child i = 2 * i + 2
+    let parent_node i = (i - 1) / 2
+
+    (* We say that a heap respects the "heap ordering" if the value of
+       each node is no greater than the value of its children. The
+       algorithm manipulates arrays that respect the heap algorithm,
+       except for one node whose value may be too small or too large.
+
+       The auxiliary functions [sift_up] and [sift_down] take
+       such a misplaced value, and move it "up" (respectively: "down")
+       until the heap ordering is restored.
+
+       Functions [sift_up] and [sift_down] do not perform swaps, but
+       rather expect the value to be assigned in the heap as an
+       additional parameter [x], resulting in twice less assignments. *)
+
+    (* store [x] at index [i], moving it up if necessary *)
+    let rec sift_up h i x =
+      if i = 0 then Dynarray.set h 0 x else
+      let p = parent_node i in
+      let y = Dynarray.get h p in
+      if E.compare x y < 0 then (
+        Dynarray.set h i y;
+        sift_up h p x
+      ) else
+        Dynarray.set h i x
+
+    let add h x =
+      let i = Dynarray.length h in
+      Dynarray.add_last h x;
+      if i > 0 then sift_up h i x
+
+    let add_seq h s =
+      Seq.iter (add h) s
+
+    exception Empty
+
+    let min_elt h =
+      if Dynarray.length h = 0 then raise Empty;
+      Dynarray.get h 0
+
+    let min_elt_opt h =
+      if Dynarray.length h = 0 then None else Some (Dynarray.get h 0)
```

Comment:
`Dynarray.is_empty` (same above for `min_elt`)

Reply:
Done.